unified-codes/tools/scripts/dgraph/migration/v1/export.js

```diff
@@ -0,0 +1,190 @@
+const outputFilename = process.argv[3] || './data/products.xlsx';
+const items = [];
+
+const parseFile = async (data) => {
+  const jsonData = JSON.parse(data);
+  const { set } = jsonData;
+  const parent = { form: '', formCategory: '', product: { code: '', description: '' } };
+
+  set.forEach((item) => parseProduct(item, parent));
+
+  return new Promise((resolve) => resolve());
+};
+
+const parseChildren = (children, parent) => {
+  if (children) {
+    children.forEach((child) => parseProduct(child, parent));
+  }
+};
+
+const getProperty = (properties, type) => {
+  if (!properties) return '';
+  const prop = properties.find((p) => p.type === type);
+  return prop ? prop.value : '';
+};
+
+const addPropertiesToProduct = (product, properties) => {
+  product.strength = getProperty(properties, 'strength');
+  product.ddd = getProperty(properties, 'ddd');
+  product.rxnav = getProperty(properties, 'code_rxnav');
+  product.eml = getProperty(properties, 'who_eml');
+  product.atc = getProperty(properties, 'code_who_atc');
+  product.unspsc = getProperty(properties, 'code_unspsc');
+  product.nzulm = getProperty(properties, 'code_nzulm');
+};
+
+const getProduct = (item, parent) => {
+  const { code, description, type } = item;
+  const product = {
+    code,
+    description,
+    type,
+    form: parent.form,
+    formCategory: parent.formCategory,
+    product: parent.product,
+  };
+  addPropertiesToProduct(product, item.has_property);
+  return product;
+};
+
+const parseProduct = (item, parent) => {
+  const { code, description, type } = item;
+  switch (item.type) {
+    case 'unit_of_use':
+      items.push(getProduct(item, parent));
+      break;
+    case 'other':
+    case 'drug':
+    case 'medicinal_product':
+      if (item.has_child) {
+        parent.product = { code, description, type };
+        addPropertiesToProduct(parent.product, item.has_property);
+        parseChildren(item.has_child, parent);
+      } else {
+        items.push(getProduct(item, parent));
+      }
+      break;
+    case 'form':
+      parent.form = description;
+      parseChildren(item.has_child, parent);
+      break;
+    case 'form_category':
+      parent.formCategory = description;
+      parseChildren(item.has_child, parent);
+      break;
+  }
+};
+
+const sortItems = (itemA, itemB) => {
+  const descriptionA = itemA.description.substring(0, 5).toUpperCase();
+  const descriptionB = itemB.description.substring(0, 5).toUpperCase();
+  if (descriptionA < descriptionB) {
+    return -1;
+  }
+  if (descriptionA > descriptionB) {
+    return 1;
+  }
+
+  const formCategoryA = itemA.formCategory;
+  const formCategoryB = itemB.formCategory;
+  if (formCategoryA < formCategoryB) {
+    return -1;
+  }
+  if (formCategoryA > formCategoryB) {
+    return 1;
+  }
+
+  const formA = itemA.form;
+  const formB = itemB.form;
+  if (formA < formB) {
+    return -1;
+  }
+  if (formA > formB) {
+    return 1;
+  }
+
+  return 0;
+};
```

Comment:
Hmm, does this work as expected? I've had some weird behaviour with multi-case sorts like this. Since the sort will use some sort of sublinear algorithm, it won't compare every item with every item (in mathsy terms, the sort function needs to be distributive, which I think explodes with multi-case comparisons). I'd expect this to work mostly, but for some items to get thrown in in random places... but if it's working, I say leave it 🤷‍♂️ .